docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,306 @@
+kind: AzureMachinePool
+metadata:
+  name: node-os-image-caching-machine-pool-machine
+  namespace: default
+  creationTimestamp: "2023-06-20T17:41:54Z"
+status:
+  conditions:
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: Ready
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: BootstrapSucceeded
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: NodeHealthy
+  latestModelApplied: true
+  ready: true
+```
+
+Example AzureMachine yaml with the important fields shown (want status.ready == true and each status: "True"):
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachine
+metadata:
+  name: node-os-image-caching-machine
+  namespace: default
+  creationTimestamp: "2023-06-20T17:38:39Z"
+status:
+  conditions:
+  - lastTransitionTime: "2023-06-20T17:41:16Z"
+    status: "True"
+    type: Ready
+  - lastTransitionTime: "2023-06-20T17:41:16Z"
+    status: "True"
+    type: BootstrapSucceeded
+  - lastTransitionTime: "2023-06-20T17:40:16Z"
+    status: "True"
+    type: DisksReady
+  - lastTransitionTime: "2023-06-20T17:40:16Z"
+    status: "True"
+    type: VMRunning
+  ready: true
+```
+
+In terms of when to take a snapshot, a day is given as a general example which should be good for typical use but the specification of how often will be customizable as we know that certain operators have different strategies and use cases for how they’re running their services on our clusters.
+
+In terms of data model changes, AzureMachineTemplate and AzureMachinePool will be changed and the changes we expect will be purely additive and nonbreaking. No removals should be required to the data model. For AzureMachineTemplate and AzureMachinePool we will add a new optional field under spec.template.image called nodePrototyping which will be enabled if present and it have a required field under it called interval which will map to an interval of 1 day or 24 hours by default.
+
+Example AzureMachineTemplate yaml:
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachineTemplate
+metadata:
+  name: node-os-image-caching-machine-template
+  namespace: default
+spec:
+  template:
+    image:
+      nodePrototyping:
+        interval: 24h
+```
+
+For AzureMachinePool, we will also add an optional field under spec.strategy.rollingUpdate called prototypeAutomaticRollout which will be set to false by default since we don't want the replacement of a node OS image to automatically trigger a rolling update of all the nodes. This is because the new image will be functionally identical to the old one outside of name itself (since the old nodes will all have the update and security patch contents already present). If for some reason the operator wants to always trigger this rollout (maybe if they programatically use the image names themselves), then they can simply set this field to true.
```

Comment:
```suggestion
For AzureMachinePool, we will also add an optional field under spec.strategy.rollingUpdate called prototypeAutomaticRollout which will be set to false by default since we don't want the replacement of a node OS image to automatically trigger a rolling update of all the nodes. This is because the new image will be functionally identical to the old one outside of name itself (since the old nodes will all have the update and security patch contents already present). If for some reason the operator wants to always trigger this rollout (maybe if they programmatically use the image names themselves), then they can simply set this field to true.
```